### PR TITLE
ARM ASM push / pop rework

### PR DESCRIPTION
**Pull request by agross-oss**
This change takes all of the ASM function calls and makes sure that we are pushing 8 byte aligned amounts onto the stack before making function calls.  The slight complication on this was v6m thumb issues with not being able to pop LR.

### CODE REVIEW

**Comment by agross-linaro** [comment]
First patch is a slightly modified version of the original at:

https://github.com/zephyrproject-rtos/zephyr/pull/12822

**Review by ioannisg — COMMENTED** [review]
Thanks for this patch @agross-linaro , it mainly looks OK, but I think we could try to solve this problem together with #12961

**Review by pizi-nordic — APPROVED** [review]

**Review by ioannisg — DISMISSED** [review]
Thanks for the patch, @agross-linaro , I left some comments inline for you to consider.

Besides those, we really need to fix what I mention in #12961. If we allow Memory Protection features on CM0+ and CM23, which it should be the case, IMO, we need to fix the push/pop in all relevant code. Overall:
- in userspace.S, there are missing cases in _arm_do_syscall, z_arch_user_string_nlen
- in swap_helper.S we need to fix the psh/pop in all occasions of CONFIG_"some MPU  function">

Perhaps we can let #12817 go in, first, because it depreciates several of these calls (by defining a single API function for MPU reprogramming?) Otherwise, we can let this one in, first, and I rebase.

File: arch/arm/core/userspace.S [hunk 1]
```diff
@@ -127,11 +127,9 @@ SECTION_FUNC(TEXT,_arm_userspace_enter)
     mov r0, lr
 
 #ifdef CONFIG_EXECUTION_BENCHMARKING
-    stm sp!,{r0-r3} /* Save regs r0 to r4 on stack */
-    push {lr}
+    push {r0, r1, r2, r3, r12, lr}
```
[the remark this answers is not in the document]
Comment (pizi-nordic) [reply arch/arm/core/userspace.S]: I assume that this code is used only in ARMv7. The pop below has LR in reglist, which suggests that. @agross-linaro Is my assumption right?

File: arch/arm/include/kernel_arch_func.h [hunk 2]
```diff
@@ -96,9 +96,9 @@ _arch_switch_to_main_thread(struct k_thread *main_thread,
 		 * entry function
 		*/
 		"mov %%r0, %3 \t\n"
-		"push {r2, lr} \t\n"
+		"push {r0, r1, r2, r3, r12, lr} \t\n"
```
[the remark this answers is not in the document]
Comment (agross-linaro) [reply arch/arm/include/kernel_arch_func.h]: we have to push in multiples of 2 if we push.  That said, regardless of our choice in the paired register, we have to handle the clobber correctly and we have to do multiples.

File: arch/arm/core/swap_helper.S [hunk 3]
```diff
@@ -45,14 +45,14 @@ SECTION_FUNC(TEXT, __pendsv)
 
 #ifdef CONFIG_TRACING
     /* Register the context switch */
-    push {lr}
+    push {r0, lr}
     bl z_sys_trace_thread_switched_out
 #if defined(CONFIG_ARMV6_M_ARMV8_M_BASELINE)
-    pop {r0}
-    mov lr, r0
+    pop {r0, r1}
```
[the remark this answers is not in the document]
Comment (agross-linaro) [reply arch/arm/core/swap_helper.S]: nope, not that I saw.

File: arch/arm/core/swap_helper.S [hunk 4]
```diff
@@ -303,19 +303,23 @@ _stack_frame_endif:
     beq _oops
 
 #if CONFIG_IRQ_OFFLOAD
-    push {lr}
+    push {r0, lr}
     blx _irq_do_offload  /* call C routine which executes the offload */
-    pop {r3}
-    mov lr, r3
-#endif
+#if defined(CONFIG_ARMV6_M_ARMV8_M_BASELINE)
```
[the remark this answers is not in the document]
Comment (agross-linaro) [reply arch/arm/core/swap_helper.S]: ah right

File: arch/arm/core/isr_wrapper.S [hunk 5]
```diff
@@ -106,28 +106,28 @@ _idle_state_cleared:
...
 	bl read_timer_end_of_isr
 #if defined(CONFIG_ARMV6_M_ARMV8_M_BASELINE)
-	pop {r3}
+	pop {r0, r3}
 	mov lr,r3
 #else
-	pop {lr}
+	pop {r0, lr}
 #endif /* CONFIG_ARMV6_M_ARMV8_M_BASELINE */
 	ldm sp!,{r0-r3} /* Restore r0 to r4 regs */
```
[the remark this answers is not in the document]
Comment (agross-linaro) [reply arch/arm/core/isr_wrapper.S]: yeah i missed fixing an originally bad comment

File: arch/arm/core/userspace.S (hunk 1 above)
[the remark this answers is not in the document]
Comment (agross-linaro) [reply arch/arm/core/userspace.S]: yeah i need to fix this.  I had changed this then thought i changed it back.  I'll fix

File: arch/arm/core/isr_wrapper.S [hunk 6]
```diff
@@ -41,7 +41,7 @@ GTEXT(_IntExit)
  */
 SECTION_FUNC(TEXT, _isr_wrapper)
 
-	push {lr}		/* lr is now the first item on the stack */
+	push {r0,lr}		/* r0, lr is now the first items on the stack */
```
[the remark this answers is not in the document]
Comment (agross-linaro) [reply arch/arm/core/isr_wrapper.S]: fixed

File: arch/arm/core/isr_wrapper.S (hunk 5 above)
[the remark this answers is not in the document]
Comment (agross-linaro) [reply arch/arm/core/isr_wrapper.S]: fixed

File: arch/arm/core/swap_helper.S (hunk 3 above)
[the remark this answers is not in the document]
Comment (agross-linaro) [reply arch/arm/core/swap_helper.S]: double checked, yeah this is at the beginning of the isr, so r0-r3 are fair game.

File: arch/arm/core/swap_helper.S [hunk 7]
```diff
@@ -242,29 +242,29 @@ _thread_irq_disabled:
...
 #if defined(CONFIG_ARMV6_M_ARMV8_M_BASELINE)
-    pop {r0}
-    mov lr, r0
+    pop {r0, r1}
+    mov lr, r1
 #else
-    pop {lr}
-#endif /* CONFIG_ARMV6_M_ARMV8_M_BASELINE */
+    pop {r0, lr}
+#endif
```
[the remark this answers is not in the document]
Comment (agross-linaro) [reply arch/arm/core/swap_helper.S]: fixed

File: arch/arm/core/swap_helper.S (hunk 4 above)
[the remark this answers is not in the document]
Comment (agross-linaro) [reply arch/arm/core/swap_helper.S]: fixed

File: arch/arm/core/swap_helper.S (hunk 7 above)
[the remark this answers is not in the document]
Comment (agross-linaro) [reply arch/arm/core/swap_helper.S]: fixed

File: arch/arm/core/swap_helper.S (hunk 4 above)
[the remark this answers is not in the document]
Comment (agross-linaro) [reply arch/arm/core/swap_helper.S]: fixed

File: arch/arm/core/userspace.S (hunk 1 above)
[the remark this answers is not in the document]
Comment (agross-linaro) [reply arch/arm/core/userspace.S]: fixed

File: arch/arm/include/kernel_arch_func.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (agross-linaro) [reply arch/arm/include/kernel_arch_func.h]: it ended up being a clobber list issue.  We can leave the push {r2,lr} and add r3 to the clobber list and all is well

**Review by ioannisg — APPROVED** [review]
Approving this, which means this could go in as is, however, we need to do more, at least in userspace.S, where there are missing cases in _arm_do_syscall, z_arch_user_string_nlen (See my earlier comment)

**Comment by ioannisg** [comment]
> 
> 
> From inspecting that code, we don't have any on bl boundaries, so we should be ok on the do_syscall and string_nlen function.

you are right

**Review by galak — APPROVED** [review]

**Pull request merged**